### PR TITLE
Functional: Scans: More General ‘Backward’ Unstructured Domain Analysis, Better Shift Analysis, Cleanups

### PR DESCRIPTION
**Pull request by fthaler**
## Description

Implements a more general ‘backward’ unstructured domain/extent analysis and proper tracing of shifts in functions etc. Further, the domain analysis part is cleaned up a bit and popping up of reductions is fixed.

## Requirements

Before submitting this PR, please make sure:

- [x] You have run the code checks, tests and documentation build successfully
- [x] All fixes and all new functionality are tested and documentation is up to date
- [x] You looked at the [review checklist](https://github.com/GridTools/gt4py/blob/master/REVIEW_CHECKLIST.txt)

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

**Comment by fthaler** [comment]
pre-commit failing because of https://github.com/GridTools/gt4py/pull/913

File: tests/functional_tests/iterator_tests/test_fvm_nabla.py [hunk 1]
```diff
@@ -351,6 +352,8 @@ def nabla_sign(n_nodes, out_MXX, out_MYY, pp, S_MXX, S_MYY, vol, node_index, is_
 
 def test_nabla_sign(fencil_processor, lift_mode):
     fencil_processor, validate = fencil_processor
+    if lift_mode != LiftMode.FORCE_INLINE:
```
Comment (havogt) [inline tests/functional_tests/iterator_tests/test_fvm_nabla.py:355]: We had a discussion yesterday about the problem of lift and concluded that it is not broken in the part where we thought it is. Let's have a chat next week.

**Review by havogt — CHANGES_REQUESTED** [review]
We probably need to make sure that this problem is resolved, otherwise we might run into strange issues in the future.

File: tests/functional_tests/iterator_tests/test_fvm_nabla.py (hunk 1 above)
Comment (havogt) [reply tests/functional_tests/iterator_tests/test_fvm_nabla.py:355]: We should resolve this separately.

**Pull request merged**